### ISSUE TITLE
[no-unused-vars] false positive for imported types when exporting namespace with same name

### ISSUE THREAD

**Issue by pgoldberg** [typescript-eslint/typescript-eslint#4594]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
```

```TS
// this interface is marked unused
interface Foo {
    type: string;
}

export namespace Foo {
    export function isFoo(obj: Foo): obj is Foo {
        return obj.type === "foo";
    }
}
```

**Expected Result**

I expect that there is no error, since the type is used in this line: `export function isFoo(obj: Foo): obj is Foo`

**Actual Result**

The `Foo` type is marked as unused.

**Additional Info**

It looks like [`isUsedVariable`](https://github.com/typescript-eslint/typescript-eslint/blob/88941065097ce71a0e385cb2e00c1b263dbc69ab/packages/eslint-plugin/src/util/collectUnusedVariables.ts#L447) is returning false for the interface, and [`variable.references`](https://github.com/typescript-eslint/typescript-eslint/blob/88941065097ce71a0e385cb2e00c1b263dbc69ab/packages/eslint-plugin/src/util/collectUnusedVariables.ts#L725) is empty for `Foo`. It seems like `no-unused-vars` thinks that `Foo` in `function isFoo(obj: Foo)` is referring to the namespace, when Typescript does correctly realize that `Foo` refers to the type in this case.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.7.0` (also tested on `5.12.1`) |
| `@typescript-eslint/parser`        | `5.7.0` (also tested on `5.12.1`) |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.20.2` |

**Comment by bradzacher** [comment]
This is really difficult because of the shadowing.
A namespace can be referenced in a type position because they can export types.
So we have to consider the possibility that `namespace Foo` could be a valid type reference.

Which means within the `namespace Foo` - you have the namespace name `Foo` and the interface name `Foo`. What TS does here is be "helpful" for you and ignore the proper `Foo` that's shadowing the outer `Foo` beacause it knows two things:
1) `Foo` exports no types, so it knows it will be an error to use it in a type location.
2) A namespace name by itself isn't valid to reference in a type location (with no `.` member access it is always a TS error).

TS can really easily do these lookups whenever it wants because it has complete type information. However, it's a lot harder for us to do these lookups.

In general this sort of name shadowing is really "smelly" code. The rules around it are complicated and often times unclear and surprising. For example:

```ts
type T1  = Foo['isFoo']; // references the interface, and errors
const x1 = Foo['isFoo']; // references the namespace, and is fine

type T2  = Foo.isFoo;    // references the namespace, and errors
const x2 = Foo.isFoo;    // references the namespace, and is fine

type T3  = Foo;          // references the interface, and is fine
const x3 = Foo;          // references the namespace, and is fine
```

Because this is a really rare edge case, is really complicated to handle, and is generally not a good code practice, I will be closing this as unfixable.
